Crash under WebCore::ShadowData::~ShadowData

cdumez opened the pull request:
#### d5d562272a83aad433001065b6e2186acc97049d
<pre>
Crash under WebCore::ShadowData::~ShadowData
<a href="https://bugs.webkit.org/show_bug.cgi?id=242609">https://bugs.webkit.org/show_bug.cgi?id=242609</a>
&lt;rdar://96588555&gt;

Reviewed by Geoffrey Garen.

We see some crashes under ~ShadowData in the wild, due to reaching the
maximum recursion depth.

The ShadowData destructor indeed destroys a whole linked list recursively
via the ShadowData::m_next data member. The length of the linked list is
under the control of the Web content so we could theoretically reach the
recursion limit (though likely due to a bug or bad content).

Without knowing exactly under what circumstances and with which content
we hit this crash, my proposal is to stop relying on recursion to destroy
this linked list.

I am also fixing operator==() as it also traverses this linked list
recursively and we see similar crashes there &lt;rdar://problem/96588560&gt;.

Test: fast/css/shadow-box-recursion-depth.html

* Source/WebCore/rendering/style/ShadowData.cpp:
(WebCore::ShadowData::deleteNextLinkedListWithoutRecursion):
* Source/WebCore/rendering/style/ShadowData.h:
(WebCore::ShadowData::~ShadowData):

Canonical link: <a href="https://commits.webkit.org/252367@main">https://commits.webkit.org/252367@main</a>
</pre>

geoffreygaren commented on Source/WebCore/rendering/style/ShadowData.cpp:
```diff
@@ -38,6 +38,15 @@ ShadowData::ShadowData(const ShadowData& o)
+    // Avoid recursion errors when the linked list is too long.
+    Vector<std::unique_ptr<ShadowData>> list;
+    list.append(WTFMove(m_next));
+    while (auto& next = list.last()->m_next)
```
I think you can do this without accumulating a Vector. Something like:

for (auto next = std::exchange(m_next, nullptr); next = std::exchange(next->m_next, nullptr)) { }

cdumez replied:
I applied the suggestion. Definitely more efficient this way.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Crash under WebCore::ShadowData::~ShadowData
https://bugs.webkit.org/show_bug.cgi?id=242609
<rdar://96588555>

Reviewed by Geoffrey Garen.

We see some crashes under ~ShadowData in the wild, due to reaching the
maximum recursion depth.

The ShadowData destructor indeed destroys a whole linked list recursively
via the ShadowData::m_next data member. The length of the linked list is
under the control of the Web content so we could theoretically reach the
recursion limit (though likely due to a bug or bad content).

Without knowing exactly under what circumstances and with which content
we hit this crash, my proposal is to stop relying on recursion to destroy
this linked list.

I am also fixing operator==() as it also traverses this linked list
recursively and we see similar crashes there <rdar://problem/96588560>.

Test: fast/css/shadow-box-recursion-depth.html

* Source/WebCore/rendering/style/ShadowData.cpp:
(WebCore::ShadowData::deleteNextLinkedListWithoutRecursion):
* Source/WebCore/rendering/style/ShadowData.h:
(WebCore::ShadowData::~ShadowData):

Canonical link: https://commits.webkit.org/252367@main
